### PR TITLE
[spirv] Translate (RW)ByteAddrressBuffer types

### PR DESCRIPTION
**Pull request by ehsannas**

### CODE REVIEW

File: tools/clang/lib/SPIRV/DeclResultIdMapper.cpp [hunk 1]
```diff
@@ -100,9 +100,17 @@ uint32_t DeclResultIdMapper::createFileVar(uint32_t varType, const VarDecl *var,
...
 uint32_t DeclResultIdMapper::createExternVar(uint32_t varType,
                                              const VarDecl *var) {
-  // TODO: storage class can also be Uniform
-  const uint32_t id = theBuilder.addModuleVar(
-      varType, spv::StorageClass::UniformConstant, var->getName(), llvm::None);
+  auto storageClass = spv::StorageClass::UniformConstant;
+
+  // TODO: Figure out other cases where the storage class should be Uniform.
+  if (auto *t = var->getType()->getAs<RecordType>()) {
+    llvm::StringRef typeName = t->getDecl()->getName();
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/DeclResultIdMapper.cpp]: can be const

File: tools/clang/lib/SPIRV/ModuleBuilder.cpp [hunk 2]
```diff
@@ -673,6 +674,52 @@ uint32_t ModuleBuilder::getSampledImageType(uint32_t imageType) {
...
+  const auto *mem0Offset = Decoration::getOffset(theContext, 0, 0);
+  const auto *byteAddressBufferType = Type::getStruct(
+      theContext, {raTypeId}, {bufferBlock, mem0NonWritable, mem0Offset});
+  const uint32_t typeId = theContext.getResultIdForType(byteAddressBufferType);
+  theModule.addType(byteAddressBufferType, typeId);
+  theModule.addDebugName(typeId, "type.ByteAddressBuffer");
+  return typeId;
+}
+
+uint32_t ModuleBuilder::getRWByteAddressBufferType() {
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/ModuleBuilder.cpp]: It seems to me that this function can be merged into the above by adding an additional parameter (e.g., `isRw`) to the above?

File: tools/clang/lib/SPIRV/TypeTranslator.cpp [hunk 3]
```diff
@@ -104,6 +104,18 @@ uint32_t TypeTranslator::translateType(QualType type) {
     return theBuilder.getMatType(vecType, rowCount);
   }
 
+  // ByteAddressBuffer and RWByteAddressBuffer types. These are RecordType and
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/TypeTranslator.cpp]: Can this be moved into `translateResourceTpye()`?

File: tools/clang/lib/SPIRV/DeclResultIdMapper.cpp (hunk 1 above)
Comment (ehsannas) [inline tools/clang/lib/SPIRV/DeclResultIdMapper.cpp]: done

File: tools/clang/lib/SPIRV/ModuleBuilder.cpp (hunk 2 above)
Comment (ehsannas) [inline tools/clang/lib/SPIRV/ModuleBuilder.cpp]: done

File: tools/clang/lib/SPIRV/TypeTranslator.cpp (hunk 3 above)
Comment (ehsannas) [inline tools/clang/lib/SPIRV/TypeTranslator.cpp]: oh, sorry I must have done this in the first place. Done.

**Pull request merged**